heudiconv/main.py

```diff
@@ -0,0 +1,340 @@
+    """
+    Overloads default sys.excepthook with our exceptionhook handler.
+
+    If interactive, our exceptionhook handler will invoke pdb.post_mortem;
+    if not interactive, then invokes default handler.
+    """
+    def _pdb_excepthook(type, value, tb):
+        if is_interactive():
+            import traceback
+            import pdb
+            traceback.print_exception(type, value, tb)
+            # print()
+            pdb.post_mortem(tb)
+        else:
+            lgr.warning(
+              "We cannot setup exception hook since not in interactive mode")
+
+    sys.excepthook = _pdb_excepthook
+
+
+def process_extra_commands(outdir, command, files, dicom_dir_template,
+                           heuristic, session, subjs, grouping):
+    """
+    Perform custom command instead of regular operations. Supported commands:
+    ['treat-json', 'ls', 'populate-templates']
+
+    Parameters
+    ----------
+    outdir : String
+        Output directory
+    args : Namespace
+        arguments
+    """
+    if command == 'treat-jsons':
+        for f in files:
+            treat_infofile(f)
+    elif command == 'ls':
+        ensure_heuristic_arg(heuristic)
+        heuristic = load_heuristic(heuristic)
+        heuristic_ls = getattr(heuristic, 'ls', None)
+        for f in files:
+            study_sessions = get_study_sessions(
+                dicom_dir_template, [f], heuristic, outdir,
+                session, subjs, grouping=grouping)
+            print(f)
+            for study_session, sequences in study_sessions.items():
+                suf = ''
+                if heuristic_ls:
+                    suf += heuristic_ls(study_session, sequences)
+                print(
+                    "\t%s %d sequences%s"
+                    % (str(study_session), len(sequences), suf)
+                )
+    elif command == 'populate-templates':
+        ensure_heuristic_arg(heuristic)
+        heuristic = load_heuristic(heuristic)
+        for f in files:
+            populate_bids_templates(f, getattr(heuristic, 'DEFAULT_FIELDS', {}))
+    elif command == 'sanitize-jsons':
+        tuneup_bids_json_files(files)
+    elif command == 'heuristics':
+        from .utils import get_known_heuristics_with_descriptions
+        for name_desc in get_known_heuristics_with_descriptions().items():
+            print("- %s: %s" % name_desc)
+    elif command == 'heuristic-info':
+        ensure_heuristic_arg(heuristic)
+        from .utils import get_heuristic_description
+        print(get_heuristic_description(heuristic, full=True))
+    else:
+        raise ValueError("Unknown command %s", command)
+    return
+
+
+def ensure_heuristic_arg(heuristic=None):
+    from .utils import get_known_heuristic_names
+    if not heuristic:
+        raise ValueError("Specify heuristic using -f. Known are: %s"
+                         % ', '.join(get_known_heuristic_names()))
+
+
+def workflow(dicom_dir_template=None, files=None, subjs=None,
```

Comment:
no sane person should invoke `workflow` providing any of those as positional args, I will add a leading `*, ` ;)

Reply:
I've seen that once, but never used it. It's awesome! I also added a numpydoc Notes section to the docstring to clarify usage for those who, like me, might be confused by the way it's defined. I can remove it if you want though.

Reply:
Notes are good, thank you!